### ISSUE TITLE
Provide better handling of front-end CSS/JS resources for caching

### ISSUE THREAD

**Issue by engram-design** [verbb/formie#968]
Currently, we serve front-end CSS and JS through Yii's asset bundles. This will produce resources like:

https://{domain}/cpresources/5ae22002/formie-theme.css
https://{domain}/cpresources/1b5a2b7b/formie.js

Which are served from the `cpresources` directory. This is great because of the ease-of use being able to use asset bundles, but this leads to a few issues.

1. Caching
Nothing should be cached under `cpresources` due to the ephemeral nature of them, as they'll be re-generated as needed. However, Formie's JS and CSS should be cached as it's served on the front-end.

2. Robots.txt
Commonly, people will have a robots.txt file with the following:

```
User-agent: *
Disallow: /cpresources/
```

While that's correct and valid (anything CP-related should be blocked) it's not for Formie.

So in summary, front-end resources are being placed in CP-related resources and they shouldn't be.

**Comment by engram-design** [comment]
So marking this as something we don't need to tackle.

1. It's safe and recommended to have web servers send cache headers for files in `cpresources`, according to @brandonkelly
2. This is only really an issue if there's HTML content in the `cpresources` folder - hence something for search engines to index. There isn't for Formie. This has been flagged by Semrush, but I suspect it's either a false positive or a misconfigured scan of a site.

**Comment by codyjames** [comment]
@engram-design My client is wanting me to fix the Semrush-related "issues" caused by this. Any recommendations?

**Comment by engram-design** [comment]
I'm not actually familiar with the issue in Semrush (was reported by another user). What's the exact error? It should only be complaining if there is HTML content in `cpresources`, which there shouldn't be. At least, Formie won't put it there.

**Comment by codyjames** [comment]
@engram-design I think the error is actually just that we are referencing assets that are blocked by Robots.txt. Here is the error description from Semrush:

"Blocked resources are resources (e.g., CSS, JavaScript, image files, etc.) that are blocked from crawling by a "Disallow" directive in your robots.txt file. By disallowing these files, you're preventing search engines from accessing them and, as a result, properly rendering and indexing your webpages. This, in return, may lead to lower rankings."

**Comment by engram-design** [comment]
Well, I guess that's technically true Semrush. But rendering a form (potentially without CSS and JS) is hardly going to affect rankings, so that's a bit of a silly statement. I see _why_ they recommend that - where a site might be JS-based and if no JS is executed you get a blank page, but that's not the case here. Formie will still render the form to search engines - it just might be unstyled, which doesn't matter for indexing.

Can you flag this as something to ignore or bypass in Semrush? I don't believe it's an actual issue in Formie's case.

**Comment by codyjames** [comment]
I feel like I could "hide" the issue for existing pages, but any new page created would make this issue appear again, I think.

I wonder if I could adjust the robots.txt to allow specifically these formie files but disallow the rest of cpresources. However, I feel like these URLs being sorta randomly generated, I'm not sure how to go about that.

**Comment by codyjames** [comment]
Also, I could hide this _general_ issue, but I'd be worried about it not surfacing other issues that are similar (but actually a problem) down the road. I've gone ahead and done that for now, so this is no longer an imminent issue for me.

**Comment by engram-design** [comment]
Yeah, the URLs will be randomly generated and changed all the time, so adding rules for those won't work. Unless you can wildcard paths like `cpresources/*/formie-theme.css` (I'm not sure if that's possible - not a `robots.txt` expert).

**Comment by khalwat** [comment]
@engram-design see here for wildcard matching rules:

https://developers.google.com/search/docs/crawling-indexing/robots/robots_txt#url-matching-based-on-path-values

TL;DR yes, wildcards are supported.

Relevant part of the RFC:

https://www.rfc-editor.org/rfc/rfc9309.html#name-special-characters

One thing I will say, though, is while there is an RFC spec for all of this, I wouldn't be shocked if some crawlers don't implement it fully.

**Comment by MattWilcox** [comment]
Just to throw in here that this is not a Formie only thing - we've had the exact same warning for the exact same reasons with Sprig, as that plugin serves its HTMX JS from inside `cpresources`, which SEMrush wasa not happy about.

Wildcarding all JS files inside `cpresources` did work for us and stopped SEMrush from throwing warnings.

Being more specific than the SEOmatic default, it wins.

```
Disallow: /cpresources/
Allow: /cpresources/*/*.js
```

**Comment by khalwat** [comment]
@MattWilcox you might want to change that to target only the SEMrush bot user agent

**Comment by MattWilcox** [comment]
Already did thanks, but for completeness sake for the thread; the relevent bit of robots.txt wants to look like this:

```
User-agent: SiteAuditBot
Allow: /cpresources/*/*.js
```

and so the SEOmatic config in the CMS would be like this:

```
# robots.txt for {{ siteUrl }}

{{ seomatic.helper.sitemapIndex() }}
{% switch seomatic.config.environment %}

{% case "live" %}

# live - don't allow web crawlers to index cpresources/ or vendor/

User-agent: *
Disallow: /cpresources/
Disallow: /vendor/
Disallow: /.env
Disallow: /cache/

# but do allow SEMrush access to the JS files so it doesn't complain
User-agent: SiteAuditBot
Allow: /cpresources/*/*.js
```

[SEMrush Reference](https://www.semrush.com/kb/681-site-audit-troubleshooting)

**Comment by lejoe** [comment]
We see that problem in google search console too:
![CleanShot 2024-12-17 at 13 29 40](https://github.com/user-attachments/assets/d60f8222-552c-4874-ac8b-cae32c32e391)

The google crawler also does screenshots / uses the rendered pages for it's calculations.
We suspect those blocked assets to prevent the good rendering of those pages.

What would be the downside of Allowing all on those resources?
